graph2tac/loader/predict_server.py

```diff
@@ -605,15 +654,20 @@ def main():
 
     log_cnts = LoggingCounters(process_uuid=process_uuid)
     model = load_model(config, log_levels)
-    predict_server = PredictServer(model, config, log_cnts)
+    response_history = ResponseHistory(recording_on=config.replay)  # only record response history if replay is on
+    predict_server = PredictServer(model, config, log_cnts, response_history)
 
     if config.record is not None:
         record_context = open(config.record, 'wb')
     else:
         record_context = contextlib.nullcontext()
 
     with record_context as record_file:
-        if not config.tcp:
+        if config.replay:
+            assert not config.tcp, "--replay not compatible with --tcp"
+            assert not config.tcp, "--replay not compatible with --record"
```

Comment:
The binary invocation for the binaries in graph2tac is already pretty complicated. It would be great if the mutual-exclusivity logic for all this stuff could be encoded into the parser instead of checked after-the-fact. That will make the `--help` information clearer. See [here](https://github.com/coq-tactician/coq-tactician-reinforce/blob/ac300ecf426f94d7a770d0a1dd08dd1e8780493b/pytact/fake_coq_client.py#L32-L42) for an example of that (and also you might look at the `argparse` docs, there are many more options.

Reply:
In order to facilitate this, and make `--replay` compatible with `--record`, I added a `record` parameter [here](https://github.com/coq-tactician/coq-tactician-reinforce/commit/707fe253cb5a1e72bba95dd7b89878135a9862cc). The relevant signature is now:
```
capnp_message_generator_from_file(message_file: BinaryIO,
                                      check = True,
                                      record: BinaryIO | None = None) -> GlobalContextMessage
```

Reply:
Great idea.  I'll do that.  Ideally I would have this as needing exactly one of `--replay`, `--stdin` and `--tcp`, but I that means adding a `--stdin` argument instead of having it be the default behavior if there is no `--replay` or `--tcp`.  @mirefek @LasseBlaauwbroek any objections?

Reply:
You can add the `--stdin` argument, and make it the default can't you? See for example https://stackoverflow.com/questions/38507675/python-argparse-mutually-exclusive-group-with-default-if-no-argument-is-given